rasterio/windows.py

```diff
@@ -350,6 +350,42 @@ def window_index(window):
     return tuple(slice(*w) for w in window)
 
 
+def round_window_to_full_tiles(window=None, block_shapes=None):
+    """
+    Round window to include full expanse of intersecting tiles.
+
+    Parameters
+    ----------
+    window : a Window or window tuple
+        The input window.
+
+    block_shapes : tuple of block shapes
+        The input raster's block shape. All bands must have the same block/stripe structure
+
+    Returns
+    -------
+    Window
+    """
+    if len(set(block_shapes)) != 1:
+        raise ValueError('All bands must have the same block/stripe structure')
+
+    height_shape = block_shapes[0][0]
+    width_shape = block_shapes[0][1]
+
+    row_range = window[0]
+    col_range = window[1]
+
+    row_min = int(row_range[0] / height_shape) * height_shape
```

Comment:
If `row_range[0]` or `col_range[0]` is negative (can be the case with a boundless read window), this will snap in the wrong direction. 

```
In [3]: tuple(round_window_to_full_tiles(((-1, 1300), (-1, 1300)), src.block_shapes))
Out[3]: ((0, 1536), (0, 1536))

In [4]: tuple(round_window_to_full_tiles(((1, 1300), (1, 1300)), src.block_shapes))
Out[4]: ((0, 1536), (0, 1536))
```